### PR TITLE
Branches - Caroline & Paige

### PR DESCRIPTION
**Pull request by stupendousC**
# Adagrams
Congratulations! You're submitting your assignment.

## Comprehension Questions

Feature | Feedback
--- | ---
What are the components that make up a method? | Signature line (inc. keyword def, method title/name, parameters), block of code, maybe a return statement, end keyword.

What are the advantages of using git when collaboratively working on one code base? | Multiple people can update & see the code.  If you mess up, you can go back to a save point.  If you want to branch out and try something new in the past, you can.

What kind of relationship did you and your pair have with the unit tests? | Great!  We can see what went wrong and figure out how to resolve the bugs.

Does your code use any methods from the `Enumerable` mixin? If so, where and why was it helpful? | .map, .each, .times, 

What was one method you and your pair used to debug code? | We used pry to fix a bug where the variable wasn't reassigning to the correct value.

What are two discussion points that you and your pair discussed when giving/receiving feedback from each other that you would be willing to share? | 1. We each see something that the other doesn't see, and that's very cool because it leads to new insights that we might not have had if we worked alone.  2.  Being upfront about our own habits and interaction style before actual coding, helps us avoid communication problems and make the work come along much more smoothly than if we had to spend time hammering things out between ourselves.

### CODE REVIEW

**Comment by stupendousC** [comment]
Oops. Incomplete answer on #4. We used enumerables in many places, it was very helpful for iterating through the data and methods like .map was helpful in reducing the line count.

**Comment by tildeee** [comment]
# Adagrams
## What We're Looking For

Feature | Feedback
--- | ---
**General** |
Answered comprehension questions | x, the table formatting broke, which is not a big deal, but just something to note :)
Small commits with meaningful commit messages | x, I would probably prefer messages that are less "Wave 2 complete" and more "implements score_word method" or "refactors score table data structure" etc
**Code Requirements** |
`draw_letters` method | x
Uses appropriate data structure to store the letter distribution | x
All tests for `draw_letters` pass | x
`uses_available_letters?` method | x
All tests for `uses_available_letters?` pass | x
`score_word` method | x
Uses appropriate data structure to store the letter scores | x
All tests for `score_word` pass | x
`highest_score_from` method | x
Appropriately handles edge cases for tie-breaking logic | x
All tests for `highest_score_from` pass | x
**Overall** |



Well done on this project, Caroline and Paige! Your code overall is extremely easy to understand because your submission has such clean, logical, readable code style and decisions. Great solutions!


I've added a few comments about small refactorings I could imagine happening if you had more time on this project.


That being said, well done overall

File: lib/adagrams.rb [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+# Assemble fresh new bag of tiles
+PILE = []
+letters.length.times do |index|
+  freq[index].times do
+    PILE << letters[index]
+  end
+end
+
+def draw_letters
```
Comment (tildeee) [inline lib/adagrams.rb:20]: Instead of the preliminary set up before the `draw_letters` method, would this work or could we change it so it would work?

File: lib/adagrams.rb [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    return false
+  end
+  
+  # check letters against letters_in_hand
+  input_array.length.times do |index|
+    if clone_letters_in_hand.include? input_array[index]
+      garbage_index = clone_letters_in_hand.rindex(input_array[index])
+      clone_letters_in_hand.delete_at(garbage_index)
+    else
+      return false
```
Comment (tildeee) [inline lib/adagrams.rb:45]: Nice!

File: lib/adagrams.rb [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+  
+  # splits input into individual capital letters 
+  input_array = input.split(//)
+  input_array = input_array.map { |letter| letter.upcase }
+  
+  if input.length > clone_letters_in_hand.length
+    return false
+  elsif input.length == 0
+    return false
+  end
```
Comment (tildeee) [inline lib/adagrams.rb:37]: Nice! Catch some edge cases before doing the hard work.

File: lib/adagrams.rb [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+end
+
+####################### WAVE 3 ##########################
+
+def score_word(word)
+  score_hash = { A: 1, E: 1, I: 1, O: 1, U: 1, L: 1, N: 1, R: 1, S: 1, T: 1, 
+    D: 2, G: 2, B: 3, C: 3, M: 3, P: 3, F: 4, H: 4, V: 4, W: 4, Y: 4, 
+    K: 5, J: 8, X: 8, Q: 10, Z: 10 }
+    
+    word_array = word.upcase.split(//)
```
Comment (tildeee) [inline lib/adagrams.rb:59]: Warning: Your indentation here is off, so it makes the rest of the file off in indentation!

File: lib/adagrams.rb [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+    
+    if word_array.length >= 7 && word_array.length < 11
+      score += 8
+    end
+    
+    return score
+  end
+  
+  ####################### WAVE 4 ##########################
+  words = []
```
Comment (tildeee) [inline lib/adagrams.rb:74]: You don't need this variable, and it's "floating" outside of any method, so feel free to delete this line

File: lib/adagrams.rb [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+        winner = word
+      elsif score == highest_score && winner.length != 10
+        if word.length == 10 || (word.length < winner.length)
+          winner = word
+        end
+      end
+    end
+    
+    return {word: winner, score: highest_score}
+  end
```
Comment (tildeee) [inline lib/adagrams.rb:94]: An elegant, clean, readable, straightforward, delicious solution for this method overall!

File: lib/adagrams.rb [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+  
+  ####################### WAVE 5 ##########################
+  
+  def is_in_english_dict?(input)
+    all_words = CSV.read('assets/dictionary-english.csv')
+    input = [input.downcase]
+    return all_words.include? input
+  end
```
Comment (tildeee) [inline lib/adagrams.rb:102]: Nice work on this optional wave! :)

Comment (tildeee) [inline lib/adagrams.rb:102]: Tests? ;)
